arrabiata/src/witness.rs

```diff
@@ -0,0 +1,1098 @@
+            self.sponge_e1[0].clone()
+        } else {
+            self.sponge_e2[0].clone()
+        };
+        let Column::X(idx) = col else {
+            unimplemented!("Only works for private columns")
+        };
+        self.state[idx] = r.clone();
+        self.r = r.clone();
+        r
+    }
+
+    unsafe fn read_sixteen_bits_chunks_folding_combiner(
+        &mut self,
+        pos: Self::Position,
+        i: u32,
+    ) -> Self::Variable {
+        let r = self.r.clone();
+        self.bitmask_be(&r, 16 * (i + 1), 16 * i, pos)
+    }
+
+    unsafe fn read_bit_of_folding_combiner(
+        &mut self,
+        pos: Self::Position,
+        i: u64,
+    ) -> Self::Variable {
+        let r = self.r.clone();
+        let bit = (r >> i) & BigInt::from(1_usize);
+        self.write_column(pos, bit.clone());
+        bit
+    }
+
+    fn load_poseidon_state(&mut self, pos: Self::Position, i: usize) -> Self::Variable {
+        assert!(
+            self.selectors[Gadget::PermutationArgument as usize][self.current_row],
+            "The permutation argument should be activated"
+        );
+        let state = if self.current_iteration % 2 == 0 {
+            self.sponge_e1[i].clone()
+        } else {
+            self.sponge_e2[i].clone()
+        };
+        self.write_column(pos, state)
+    }
+
+    fn get_poseidon_round_constant(
+        &mut self,
+        pos: Self::Position,
+        round: usize,
+        i: usize,
+    ) -> Self::Variable {
+        let rc = if self.current_iteration % 2 == 0 {
+            poseidon_3_60_0_5_5_fp::static_params().round_constants[round][i]
+                .to_biguint()
+                .into()
+        } else {
+            poseidon_3_60_0_5_5_fq::static_params().round_constants[round][i]
+                .to_biguint()
+                .into()
+        };
+        self.write_public_input(pos, rc)
+    }
+
+    fn get_poseidon_mds_matrix(&mut self, i: usize, j: usize) -> Self::Variable {
+        if self.current_iteration % 2 == 0 {
+            poseidon_3_60_0_5_5_fp::static_params().mds[i][j]
+                .to_biguint()
+                .into()
+        } else {
+            poseidon_3_60_0_5_5_fq::static_params().mds[i][j]
+                .to_biguint()
+                .into()
+        }
+    }
+
+    unsafe fn save_poseidon_state(&mut self, x: Self::Variable, i: usize) {
+        assert!(
+            self.selectors[Gadget::PermutationArgument as usize][self.current_row],
+            "The permutation argument should be activated"
+        );
+        if self.current_iteration % 2 == 0 {
+            let modulus: BigInt = Fp::modulus_biguint().into();
+            self.sponge_e1[i] = x.mod_floor(&modulus)
+        } else {
+            let modulus: BigInt = Fq::modulus_biguint().into();
+            self.sponge_e2[i] = x.mod_floor(&modulus)
+        }
+    }
+
+    // The following values are expected to be absorbed in order:
+    // - vk
```

Comment:
Not sure we need to abosrb the vk, let's talk about this offline

Reply:
You're right. We do not need. Leftover of Nova reading.

Reply:
Addressed in 9452aa6be30ea23751930b575d69743ba6fb5f1a